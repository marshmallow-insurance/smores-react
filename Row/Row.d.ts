import { FC } from 'react';
import { Color } from '../theme';
import { MarginProps } from '../utils/space';
export type RowProps = {
    iconLeft?: string;
    iconLeftColor?: Color;
    iconRight?: string;
    iconRightColor?: Color;
    handleClick?: () => void;
    heading: string;
    subHeading?: string;
    type?: 'first' | 'last' | 'curved';
    borderTop?: boolean;
    borderBottom?: boolean;
    boldHeading?: boolean;
} & MarginProps;
export declare const Row: FC<RowProps>;
