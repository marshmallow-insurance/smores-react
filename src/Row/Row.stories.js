import React from 'react'
import { Row } from '../Row'

export default {
  title: 'Row',
  conmponent: Row,
}

const Template = (args) => <Row {...args} />

export const Default = Template.bind({})

Default.args = {
  icon: 'info',
  heading: 'A lovely heading',
  subHeading: 'The quick brown fox jumps over the lazy dog',
  caret: true,
  type: 'curved',
  borderTop: true,
  borderBottom: true,
}

export const NoIcon = Template.bind({})

NoIcon.args = {
  heading: 'A lovely heading',
  subHeading: 'A row without an Icon',
}

export const NoSubheading = Template.bind({})

NoSubheading.args = {
  heading: 'A lovely heading without a subheading',
  icon: 'info',
}

export const NoCaret = Template.bind({})

NoCaret.args = {
  heading: 'A lovely heading',
  subHeading: 'A row without a caret',
  icon: 'info',
  caret: false,
}

export const First = Template.bind({})

First.args = {
  heading: 'The first row',
  subHeading: 'A first row with curved corners at the top and no bottom border',
  type: 'first',
  borderBottom: false,
}

export const Middle = Template.bind({})

Middle.args = {
  heading: 'A middle row',
  subHeading: `A row with angular corners to go between 'first' and 'last' rows`,
}

export const Last = Template.bind({})

Last.args = {
  heading: 'The last row',
  subHeading: 'A last row with curved corners at the bottom and no top border',
  type: 'last',
  borderTop: false,
}

export const Curved = Template.bind({})

Curved.args = {
  heading: 'A curved row',
  type: 'curved',
  subHeading: 'A curved row with a border radius of 8px on all corners',
}
