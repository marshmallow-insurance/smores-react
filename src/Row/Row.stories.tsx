import React from 'react'
import { Row, RowProps } from './Row'

export default {
  title: 'Row',
  component: Row,
  argTypes: { handleClick: { action: 'clicked' } },
}

const Template = (props: RowProps) => <Row {...props} />

export const Default = Template.bind({})

Default.args = {
  iconLeft: 'info',
  iconLeftColor: 'liquorice',
  iconRight: 'caret',
  iconRightColor: 'subtext',
  heading: 'A lovely heading',
  subHeading: 'The quick brown fox jumps over the lazy dog',
  type: 'curved',
  borderTop: true,
  borderBottom: true,
}

export const NoIcons = Template.bind({})

NoIcons.args = {
  heading: 'A lovely heading',
  subHeading: 'A row without an IconLeft or IconRight',
}

export const NoSubheading = Template.bind({})

NoSubheading.args = {
  heading: 'A lovely heading without a subheading',
  iconLeft: 'info',
  iconRight: 'download',
}

export const BoldHeading = Template.bind({})

BoldHeading.args = {
  heading: 'A row with a bold heading',
  subHeading: `A row with a heading that has a font weight of 500`,
  boldHeading: true,
}

export const First = Template.bind({})

First.args = {
  heading: 'The first row',
  subHeading: 'A first row with curved corners at the top and no bottom border',
  type: 'first',
  borderBottom: false,
}

export const Middle = Template.bind({})

Middle.args = {
  heading: 'A middle row',
  subHeading: `A row with angular corners to go between 'first' and 'last' rows`,
}

export const Last = Template.bind({})

Last.args = {
  heading: 'The last row',
  subHeading: 'A last row with curved corners at the bottom and no top border',
  type: 'last',
  borderTop: false,
}

export const Curved = Template.bind({})

Curved.args = {
  heading: 'A curved row',
  type: 'curved',
  subHeading: 'A curved row with a border radius of 8px on all corners',
}
