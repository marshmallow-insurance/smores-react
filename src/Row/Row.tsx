import React, { FC } from 'react'
import styled, { css } from 'styled-components'

import { Box } from '../Box'
import { Icon } from '../Icon'
import { Icons } from '../Icon/iconsList'
import { Text } from '../Text'
import { Color, theme } from '../theme'
import { MarginProps } from '../utils/space'

export type RowProps = {
  iconLeft?: Icons
  iconLeftColor?: Color
  iconRight?: Icons
  iconRightColor?: Color
  handleClick?: () => void
  heading: string
  subHeading?: string
  type?: 'first' | 'last' | 'curved'
  borderTop?: boolean
  borderBottom?: boolean
  boldHeading?: boolean
} & MarginProps

export const Row: FC<RowProps> = ({
  iconLeft,
  iconLeftColor = 'liquorice',
  iconRight,
  iconRightColor = 'marzipan',
  handleClick,
  heading,
  subHeading,
  type,
  borderTop = true,
  borderBottom = true,
  boldHeading,
  ...marginProps
}) => {
  const windowWidth = screen.width

  return (
    <Container
      type={type}
      iconLeft={iconLeft}
      borderTop={borderTop}
      borderBottom={borderBottom}
      iconRight={iconRight}
      onClick={handleClick}
      boldHeading={boldHeading}
      {...marginProps}
    >
      {iconLeft && (
        <Icon
          render={iconLeft}
          size={windowWidth > 768 ? 24 : 18}
          color={iconLeftColor}
        />
      )}
      <Box>
        <Text tag="h1" typo="base">
          {heading}
        </Text>
        <Text tag="p" typo="desc-small" color="sesame">
          {subHeading}
        </Text>
      </Box>
      {iconRight && (
        <Icon
          className="iconRight"
          render={iconRight}
          size={24}
          rotate={iconRight === 'caret' ? -90 : 0}
          color={iconRightColor}
        />
      )}
    </Container>
  )
}

interface IContainer {
  type?: 'first' | 'last' | 'curved'
  iconLeft?: string
  iconRight?: string
  borderTop: boolean
  borderBottom: boolean
  boldHeading?: boolean
}

const Container = styled(Box)<IContainer>(
  ({ type, iconLeft, borderTop, borderBottom, boldHeading }) => css`
    border-radius: ${(type === 'first' && `16px 16px 0 0`) ||
    (type === 'curved' && `16px`) ||
    (type === 'last' && `0 0 16px 16px`) ||
    0};

    border-top: ${borderTop && type !== 'curved' && type !== 'first'
      ? `1px solid ${theme.colors.oatmeal}`
      : 'none'};
    border-bottom: ${borderBottom && type !== 'curved' && type !== 'last'
      ? `1px solid ${theme.colors.oatmeal}`
      : 'none'};

    background-color: ${theme.colors.custard};
    display: grid;
    grid-template-columns: ${iconLeft ? `24px auto 24px` : `auto 24px`};
    padding: 16px;
    grid-gap: 16px;
    align-items: center;

    cursor: pointer;

    h1 {
      font-weight: ${boldHeading && 500};
    }

    .iconRight {
      justify-self: end;
    }

    @media (min-width: 768px) {
      grid-gap: 24px;
    }
  `,
)
