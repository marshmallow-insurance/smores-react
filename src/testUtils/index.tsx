import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
