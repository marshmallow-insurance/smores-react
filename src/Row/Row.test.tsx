import React from 'react'
import { screen, render } from '../testUtils'
import { Row } from './Row'

describe('Row', () => {
  describe('headings', () => {
    it('renders heading', () => {
      const { baseElement } = render(<Row heading="A lovely heading" />)

      expect(baseElement).toMatchSnapshot()

      expect(screen.getByText('A lovely heading')).toBeTruthy()
    })

    it('renders bold heading', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" boldHeading />,
      )

      expect(baseElement).toMatchSnapshot()

      expect(screen.getByText('A lovely heading')).toBeTruthy()
    })

    it('renders subHeading', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" subHeading="a nice sub heading" />,
      )

      expect(baseElement).toMatchSnapshot()

      expect(screen.getByText('a nice sub heading')).toBeTruthy()
    })
  })

  describe('type', () => {
    it('renders first type', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" type="first" />,
      )

      expect(baseElement).toMatchSnapshot()
    })

    it('renders curved type', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" type="curved" />,
      )

      expect(baseElement).toMatchSnapshot()
    })

    it('renders curved last', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" type="last" />,
      )

      expect(baseElement).toMatchSnapshot()
    })
  })

  describe('icons', () => {
    it('renders iconLeft', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" iconLeft="info" />,
      )

      expect(baseElement).toMatchSnapshot()
    })
    it('renders iconRight', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" iconRight="info" />,
      )

      expect(baseElement).toMatchSnapshot()
    })
  })

  describe('borders', () => {
    it('renders borderTop', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" borderTop borderBottom={false} />,
      )

      expect(baseElement).toMatchSnapshot()
    })
    it('renders borderBottom', () => {
      const { baseElement } = render(
        <Row heading="A lovely heading" borderTop={false} borderBottom />,
      )

      expect(baseElement).toMatchSnapshot()
    })
  })
})
