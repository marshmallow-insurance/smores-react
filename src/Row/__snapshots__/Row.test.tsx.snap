// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Row > borders > renders borderBottom 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe jrLFzT"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > borders > renders borderTop 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe joyHXT"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > headings > renders bold heading 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe itoHSn"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > headings > renders heading 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe cbpIdC"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > headings > renders subHeading 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe cbpIdC"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        >
          a nice sub heading
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > icons > renders iconLeft 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe iQFqtA"
    >
      <span
        class="sc-bRKDuR bHjpxP sc-fhHczv lfpoiQ"
        color="liquorice"
        data-testid="info-container"
        rotate="0"
        size="18"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 0C5.383 0 0 5.383 0 12s5.383 12 12 12 12-5.383 12-12S18.617 0 12 0zm-1 6.99a1 1 0 112 0c0 .553-.447 1.005-1 1.005-.553 0-1-.442-1-.995v-.01zM11 11a1 1 0 112 0v7a1 1 0 11-2 0v-7zm-9 1c0 5.514 4.486 10 10 10s10-4.486 10-10S17.514 2 12 2 2 6.486 2 12z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > icons > renders iconRight 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe cbpIdC"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
      <span
        class="sc-bRKDuR bHjpxP sc-fhHczv cTkdi iconRight"
        color="marzipan"
        data-testid="info-container"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 0C5.383 0 0 5.383 0 12s5.383 12 12 12 12-5.383 12-12S18.617 0 12 0zm-1 6.99a1 1 0 112 0c0 .553-.447 1.005-1 1.005-.553 0-1-.442-1-.995v-.01zM11 11a1 1 0 112 0v7a1 1 0 11-2 0v-7zm-9 1c0 5.514 4.486 10 10 10s10-4.486 10-10S17.514 2 12 2 2 6.486 2 12z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;

exports[`Row > type > renders curved last 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe TJUoj"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > type > renders curved type 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe dDYruh"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Row > type > renders first type 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jwTyAe cfQTIj"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <h1
          class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
          cursor="inherit"
          title=""
        >
          A lovely heading
        </h1>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ hZFdzh"
          cursor="inherit"
          title=""
        />
      </div>
    </div>
  </div>
</body>
`;
